src/main/kotlin/org/rust/lang/core/types/ty/TyTypeParameter.kt

```diff
@@ -34,7 +36,27 @@ class TyTypeParameter private constructor(
         bounds.flatMap { it.flattenHierarchy }
 
     override fun canUnifyWith(other: Ty, project: Project, mapping: TypeMapping?): Boolean {
-        mapping?.merge(mutableMapOf(this to other))
+        if (mapping != null) {
```

Comment:
Usually, it's better to avoid deep nesting and deal with short easy cases upfront:

```
if (mapping == null) return true
// Kotlin undestands that `mapping` is not null here
```